### ISSUE TITLE
NextJS Turbopack fails to compile

### ISSUE THREAD

**Issue by Anyatsu** [nrwl/nx#26510]
### Current Behavior

Running a NextJS app with the --turbo flag fails to compile when entering any page

### Expected Behavior

The --turbo flag should work

### Steps to Reproduce

1. run npx create-nx-workspace@latest
2. run development with --turbo flag

### Nx Report

```shell
NX   Report complete - copy this into the issue template

Node   : 20.12.0
OS     : linux-x64
npm    : 10.5.0

nx                 : 19.2.3
@nx/js             : 19.2.3
@nx/jest           : 19.2.3
@nx/linter         : 19.2.3
@nx/eslint         : 19.2.3
@nx/workspace      : 19.2.3
@nx/devkit         : 19.2.3
@nx/eslint-plugin  : 19.2.3
@nx/next           : 19.2.3
@nx/react          : 19.2.3
@nrwl/tao          : 19.2.3
@nx/web            : 19.2.3
@nx/webpack        : 19.2.3
typescript         : 5.4.5
---------------------------------------
Registered Plugins:
@nx/next/plugin
@nx/jest/plugin
@nx/eslint/plugin
```

### Failure Logs

```shell
> nx run nx-turbo:dev --turbo

> next dev --turbo

  ▲ Next.js 14.2.3 (turbo)
  - Local:        http://localhost:3000

 ✓ Starting...
 ✓ Ready in 2.2s
 ○ Compiling / ...
 ⨯ ./node_modules/nx/src/project-graph/plugins/isolation
Module not found: Can't resolve '/ROOT/node_modules/nx/src/project-graph/plugins/isolation/plugin-worker'
server relative imports are not implemented yet. Please try an import relative to the file you are importing from.


https://nextjs.org/docs/messages/module-not-found


 ⨯ ./node_modules/nx/src/project-graph/plugins/isolation/plugin-pool.js:27:20
Module not found: Can't resolve '/ROOT/node_modules/nx/src/project-graph/plugins/isolation/plugin-worker'
  25 |             : {}),
  26 |     };
> 27 |     const worker = (0, child_process_1.fork)(workerPath, [], {
     |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 28 |         stdio: ['ignore', 'inherit', 'inherit', 'ipc'],
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 29 |         env,
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 30 |         execArgv: [
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 31 |             ...process.execArgv,
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 32 |             // If the worker is typescript, we need to register ts-node
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 33 |             ...(isWorkerTypescript ? ['-r', 'ts-node/register'] : []),
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 34 |         ],
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 35 |     });
     | ^^^^^^^
  36 |     worker.send({ type: 'load', payload: { plugin, root } });
  37 |     // logger.verbose(`[plugin-worker] started worker: ${worker.pid}`);
  38 |     const pendingPromises = new Map();



https://nextjs.org/docs/messages/module-not-found


 ⨯ ./node_modules/nx/src/tasks-runner
Module not found: Can't resolve '/ROOT/node_modules/nx/src/tasks-runner/batch/run-batch.js'
server relative imports are not implemented yet. Please try an import relative to the file you are importing from.


https://nextjs.org/docs/messages/module-not-found


 ⨯ ./node_modules/nx/src/tasks-runner
Module not found: Can't resolve '/ROOT/node_modules/nx/src/tasks-runner/remove-old-cache-records.js'
server relative imports are not implemented yet. Please try an import relative to the file you are importing from.


https://nextjs.org/docs/messages/module-not-found


 ⨯ ./node_modules/@nx/devkit/src/utils/convert-nx-executor.js:51:12
Module not found: Can't resolve '@angular-devkit/architect'
  49 |         return toObservable(promise());
  50 |     };
> 51 |     return require('@angular-devkit/architect').createBuilder(builderFunction);
     |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  52 | }
  53 | exports.convertNxExecutor = convertNxExecutor;
  54 | function toObservable(promiseOrAsyncIterator) {



https://nextjs.org/docs/messages/module-not-found


 ⨯ ./node_modules/nx/src/adapter/ngcli-adapter.js:109:27
Module not found: Can't resolve '@angular-devkit/architect'
  107 | exports.createBuilderContext = createBuilderContext;
  108 | async function scheduleTarget(root, opts, verbose) {
> 109 |     const { Architect } = require('@angular-devkit/architect');
      |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  110 |     const logger = (0, exports.getLogger)(verbose);
  111 |     const fsHost = new NxScopedHostForBuilders(root);
  112 |     const { workspace } = await core_1.workspaces.readWorkspace('angular.json', core_1.workspaces.createWorkspaceHost(fsHost));



https://nextjs.org/docs/messages/module-not-found


 ⨯ ./node_modules/nx/src/adapter/ngcli-adapter.js:38:27
Module not found: Can't resolve '@angular-devkit/architect'
  36 |         return [];
  37 |     });
> 38 |     const { Architect } = require('@angular-devkit/architect');
     |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  39 |     const architect = new Architect(architectHost, registry);
  40 |     const { firstValueFrom } = require('rxjs');
  41 |     const toPromise = (obs) => firstValueFrom ? firstValueFrom(obs) : obs.toPromise();



https://nextjs.org/docs/messages/module-not-found


 ⨯ ./node_modules/nx/src/adapter/ngcli-adapter.js:782:129
Module not found: Can't resolve '@angular-devkit/architect/node'
  780 | }
  781 | async function getWrappedWorkspaceNodeModulesArchitectHost(workspace, root, projects) {
> 782 |     const { WorkspaceNodeModulesArchitectHost: AngularWorkspaceNodeModulesArchitectHost, } = await Promise.resolve().then(() => require('@angular-devkit/architect/node'));
      |                                                                                                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  783 |     class WrappedWorkspaceNodeModulesArchitectHost extends AngularWorkspaceNodeModulesArchitectHost {
  784 |         constructor(workspace, root, projects) {
  785 |             super(workspace, root);



https://nextjs.org/docs/messages/module-not-found


 ⨯ ./node_modules/nx/src/adapter/ngcli-adapter.js:4:16
Module not found: Can't resolve '@angular-devkit/core'
  2 | Object.defineProperty(exports, "__esModule", { value: true });
  3 | exports.getLogger = exports.wrapAngularDevkitSchematic = exports.mockSchematicsForTesting = exports.runMigration = exports.generate = exports.NxScopeHostUsedForWrappedSchematics = exports.arrayBufferToString = exports.NxScopedHostForBuilders = exports.NxScopedHost = exports.scheduleTarget = exports.createBuilderContext = void 0;
> 4 | const core_1 = require("@angular-devkit/core");
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  5 | const node_1 = require("@angular-devkit/core/node");
  6 | const chalk = require("chalk");
  7 | const path_1 = require("path");



https://nextjs.org/docs/messages/module-not-found


 ⨯ ./node_modules/nx/src/adapter/ngcli-adapter.js:29:28
Module not found: Can't resolve '@angular-devkit/core'
  27 |     // the top level import is not patched because it is imported before the
  28 |     // patching happens so we require it here to use the patched version below
> 29 |     const { workspaces } = require('@angular-devkit/core');
     |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  30 |     const { workspace } = await workspaces.readWorkspace('angular.json', workspaces.createWorkspaceHost(fsHost));
  31 |     const architectHost = await getWrappedWorkspaceNodeModulesArchitectHost(workspace, context.root, context.projectsConfigurations.projects);
  32 |     const registry = new core_1.schema.CoreSchemaRegistry();



https://nextjs.org/docs/messages/module-not-found


 ⨯ ./node_modules/nx/src/adapter/ngcli-adapter.js:5:16
Module not found: Can't resolve '@angular-devkit/core/node'
  3 | exports.getLogger = exports.wrapAngularDevkitSchematic = exports.mockSchematicsForTesting = exports.runMigration = exports.generate = exports.NxScopeHostUsedForWrappedSchematics = exports.arrayBufferToString = exports.NxScopedHostForBuilders = exports.NxScopedHost = exports.scheduleTarget = exports.createBuilderContext = void 0;
  4 | const core_1 = require("@angular-devkit/core");
> 5 | const node_1 = require("@angular-devkit/core/node");
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  6 | const chalk = require("chalk");
  7 | const path_1 = require("path");
  8 | const rxjs_1 = require("rxjs");



https://nextjs.org/docs/messages/module-not-found


 ⨯ ./node_modules/nx/src/adapter/ngcli-adapter.js:171:56
Module not found: Can't resolve '@angular-devkit/schematics'
  169 |         packageManager: (0, package_manager_1.detectPackageManager)(),
  170 |         root: (0, core_1.normalize)(root),
> 171 |         registry: new core_1.schema.CoreSchemaRegistry(require('@angular-devkit/schematics').formats.standardFormats),
      |                                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  172 |         resolvePaths: [process.cwd(), root],
  173 |         engineHostCreator: (options) => createNodeModulesEngineHost(options.resolvePaths, projects),
  174 |     });



https://nextjs.org/docs/messages/module-not-found


 ⨯ ./node_modules/nx/src/adapter/ngcli-adapter.js:176:50
Module not found: Can't resolve '@angular-devkit/schematics/tools'
  174 |     });
  175 |     workflow.registry.addPostTransform(core_1.schema.transforms.addUndefinedDefaults);
> 176 |     workflow.engineHost.registerOptionsTransform(require('@angular-devkit/schematics/tools').validateOptionsWithSchema(workflow.registry));
      |                                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  177 |     if (opts.interactive) {
  178 |         workflow.registry.usePromptProvider(createPromptProvider());
  179 |     }



https://nextjs.org/docs/messages/module-not-found


 ⨯ ./node_modules/nx/src/adapter/ngcli-adapter.js:165:26
Module not found: Can't resolve '@angular-devkit/schematics/tools'
  163 | }
  164 | function createWorkflow(fsHost, root, opts, projects) {
> 165 |     const NodeWorkflow = require('@angular-devkit/schematics/tools').NodeWorkflow;
      |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  166 |     const workflow = new NodeWorkflow(fsHost, {
  167 |         force: false,
  168 |         dryRun: opts.dryRun,



https://nextjs.org/docs/messages/module-not-found


 ⨯ ./node_modules/nx/src/adapter/ngcli-adapter.js:134:35
Module not found: Can't resolve '@angular-devkit/schematics/tools'
  132 | exports.scheduleTarget = scheduleTarget;
  133 | function createNodeModulesEngineHost(resolvePaths, projects) {
> 134 |     const NodeModulesEngineHost = require('@angular-devkit/schematics/tools')
      |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  135 |         .NodeModulesEngineHost;
  136 |     class NxNodeModulesEngineHost extends NodeModulesEngineHost {
  137 |         constructor() {



https://nextjs.org/docs/messages/module-not-found


 ⨯ ./node_modules/@swc/core/index.js:68:28
Module not found: Can't resolve '@swc/wasm'
  66 |     catch (_) {
  67 |         // postinstall supposed to install `@swc/wasm` already
> 68 |         fallbackBindings = require("@swc/wasm");
     |                            ^^^^^^^^^^^^^^^^^^^^
  69 |     }
  70 |     finally {
  71 |         return binding;



https://nextjs.org/docs/messages/module-not-found


 ⨯ ./node_modules/prettier/third-party.js:83:34
Module not found
  81 |       delete require.cache[filePath];
  82 |       const parent = require.cache[parentPath];
> 83 |       return parent === void 0 ? require(filePath) : parent.require(filePath);
     |                                  ^^^^^^^^^^^^^^^^^
  84 |     };
  85 |   }
  86 | });



https://nextjs.org/docs/messages/module-not-found


 ⨯ ./node_modules/prettier/index.js:38143:10
Module not found
  38141 |       const externalPlugins = [...uniqByKey([...externalManualLoadPluginInfos, ...externalAutoLoadPluginInfos], "requirePath").map((externalPluginInfo) => Object.assign({
  38142 |         name: externalPluginInfo.name
> 38143 |       }, require(externalPluginInfo.requirePath))), ...externalPluginInstances];
        |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  38144 |       return [...internalPlugins, ...externalPlugins];
  38145 |     }
  38146 |     function findPluginsInNodeModules(nodeModulesDir) {



https://nextjs.org/docs/messages/module-not-found


 ⨯ ./node_modules/nx/src/adapter/ngcli-adapter.js:588:26
Module not found
  586 |     }
  587 |     if ((0, path_1.extname)(name)) {
> 588 |         collectionPath = require.resolve(name);
      |                          ^^^^^^^^^^^^^^^^^^^^^
  589 |     }
  590 |     else {
  591 |         const { path: packageJsonPath, packageJson } = (0, package_json_1.readModulePackageJson)(name, (0, installation_directory_1.getNxRequirePaths)(process.cwd()));



https://nextjs.org/docs/messages/module-not-found


 ⨯ ./node_modules/yargs-parser/build/index.cjs:1029:20
Module not found
  1027 |     require: (path) => {
  1028 |         if (typeof require !== 'undefined') {
> 1029 |             return require(path);
       |                    ^^^^^^^^^^^^^
  1030 |         }
  1031 |         else if (path.match(/\.json$/)) {
  1032 |             return JSON.parse(fs.readFileSync(path, 'utf8'));



https://nextjs.org/docs/messages/module-not-found


 ⨯ ./node_modules/nx/src/adapter/ngcli-adapter.js:153:38
Module not found
  151 |                     throw new Error(`The "${name}" package does not support Nx generators or Angular Devkit schematics.`);
  152 |                 }
> 153 |                 collectionFilePath = require.resolve((0, path_1.join)((0, path_1.dirname)(packageJsonPath), schematics ?? generators));
      |                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  154 |             }
  155 |             return collectionFilePath;
  156 |         }



https://nextjs.org/docs/messages/module-not-found


 ⨯ ./node_modules/nx/src/native/index.js:53:28
Module not found
  51 |     localNodeFiles.some((f) => modulePath.endsWith(f))
  52 |   ) {
> 53 |     const nativeLocation = require.resolve(modulePath);
     |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  54 |     const fileName = basename(nativeLocation);
  55 |
  56 |     // we copy the file to a workspace-scoped tmp directory and prefix with nxVersion to avoid stale files being loaded



https://nextjs.org/docs/messages/module-not-found


 ⨯ ./node_modules/tsconfig-paths/lib/filesystem.js:31:12
Module not found
  29 |     }
  30 |     // eslint-disable-next-line @typescript-eslint/no-require-imports
> 31 |     return require(packageJsonPath);
     |            ^^^^^^^^^^^^^^^^^^^^^^^^
  32 | }
  33 | exports.readJsonFromDiskSync = readJsonFromDiskSync;
  34 | function readJsonFromDiskAsync(path,



https://nextjs.org/docs/messages/module-not-found


 ⨯ ./node_modules/ts-node/dist-raw/node-internal-modules-cjs-helpers.js:67:21
Module not found
  65 |       get: () => {
  66 |         // Node 12 hack; remove when we drop node12 support
> 67 |         const lib = (dummyModule.require || require)(name);
     |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  68 |
  69 |         // Disable the current getter/setter and set up a new
  70 |         // non-enumerable property.



https://nextjs.org/docs/messages/module-not-found


 ⨯ ./node_modules/nx/src/adapter/ngcli-adapter.js:808:39
Module not found
  806 |                 throw new Error(`The "${nodeModule}" package does not support Nx executors or Angular Devkit Builders.`);
  807 |             }
> 808 |             const executorsFilePath = require.resolve((0, path_1.join)((0, path_1.dirname)(packageJsonPath), executorsFile));
      |                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  809 |             const executorsJson = (0, fileutils_1.readJsonFile)(executorsFilePath);
  810 |             const executorConfig = executorsJson.builders?.[builder] ?? executorsJson.executors?.[builder];
  811 |             if (!executorConfig) {



https://nextjs.org/docs/messages/module-not-found


 ⨯ ./node_modules/nx/src/command-line/run/executor-utils.js:65:31
Module not found
  63 |         throw new Error(`The "${nodeModule}" package does not support Nx executors.`);
  64 |     }
> 65 |     const executorsFilePath = require.resolve((0, path_1.join)((0, path_1.dirname)(packageJsonPath), executorsFile));
     |                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  66 |     const executorsJson = (0, fileutils_1.readJsonFile)(executorsFilePath);
  67 |     const executorConfig = executorsJson.executors?.[executor] || executorsJson.builders?.[executor];
  68 |     if (!executorConfig) {



https://nextjs.org/docs/messages/module-not-found


 ⨯ ./node_modules/nx/src/project-graph/plugins/loader.js:212:64
Module not found
  210 | exports.loadNxPluginAsync = loadNxPluginAsync;
  211 | async function importPluginModule(pluginPath) {
> 212 |     const m = await Promise.resolve(`${pluginPath}`).then(s => require(s));
      |                                                                ^^^^^^^^^^
  213 |     if (m.default &&
  214 |         ('createNodes' in m.default || 'createDependencies' in m.default)) {
  215 |         return m.default;



https://nextjs.org/docs/messages/module-not-found


 ⨯ ./node_modules/prettier/index.js:18428:31
Module not found
  18426 |                 paths: [dir]
  18427 |               });
> 18428 |               result.config = require(modulePath);
        |                               ^^^^^^^^^^^^^^^^^^^
  18429 |             }
  18430 |             if (typeof result.config !== "object") {
  18431 |               throw new TypeError(`Config is only allowed to be an object, but received ${typeof result.config} in "${result.filepath}"`);



https://nextjs.org/docs/messages/module-not-found


 ⨯ ./node_modules/ts-node/dist/transpilers/swc.js:12:23
Module not found
  10 |     if (typeof swc === 'string') {
  11 |         swcDepName = swc;
> 12 |         swcInstance = require(transpilerConfigLocalResolveHelper(swc, true));
     |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  13 |     }
  14 |     else if (swc == null) {
  15 |         let swcResolved;



https://nextjs.org/docs/messages/module-not-found


 ⨯ ./node_modules/ts-node/dist/index.js:247:39
Module not found
  245 |                 : projectLocalResolveHelper;
  246 |             const transpilerPath = transpilerConfigLocalResolveHelper(transpilerName, true);
> 247 |             const transpilerFactory = require(transpilerPath)
      |                                       ^^^^^^^^^^^^^^^^^^^^^^^
  248 |                 .create;
  249 |             return createTranspiler;
  250 |             function createTranspiler(compilerOptions, nodeModuleEmitKind) {



https://nextjs.org/docs/messages/module-not-found


 ⨯ ./node_modules/ts-node/dist/transpilers/swc.js:29:23
Module not found
  27 |             }
  28 |         }
> 29 |         swcInstance = require(swcResolved);
     |                       ^^^^^^^^^^^^^^^^^^^^
  30 |     }
  31 |     else {
  32 |         swcInstance = swc;



https://nextjs.org/docs/messages/module-not-found


 ⨯ ./node_modules/@nx/devkit/src/utils/package-json.js:370:12
Module not found
  368 | }
  369 | function getPackageVersion(pkg) {
> 370 |     return require((0, path_1.join)(pkg, 'package.json')).version;
      |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  371 | }
  372 | /**
  373 |  * @description The version of Nx used by the workspace. Returns null if no version is found.



https://nextjs.org/docs/messages/module-not-found


 ⨯ ./node_modules/nx/src/config/schema-utils.js:20:24
Module not found
  18 |             (0, plugins_1.registerPluginTSTranspiler)();
  19 |         }
> 20 |         const module = require(modulePath);
     |                        ^^^^^^^^^^^^^^^^^^^
  21 |         return implementationExportName
  22 |             ? module[implementationExportName]
  23 |             : module.default ?? module;



https://nextjs.org/docs/messages/module-not-found


 ✓ Compiled / in 29.2s
 ⨯ ModuleBuildError: ./node_modules/nx/src/project-graph/plugins/isolation
Module not found: Can't resolve '/ROOT/node_modules/nx/src/project-graph/plugins/isolation/plugin-worker'
server relative imports are not implemented yet. Please try an import relative to the file you are importing from.


https://nextjs.org/docs/messages/module-not-found



./node_modules/nx/src/project-graph/plugins/isolation/plugin-pool.js:27:20
Module not found: Can't resolve '/ROOT/node_modules/nx/src/project-graph/plugins/isolation/plugin-worker'
  25 |             : {}),
  26 |     };
> 27 |     const worker = (0, child_process_1.fork)(workerPath, [], {
     |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 28 |         stdio: ['ignore', 'inherit', 'inherit', 'ipc'],
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 29 |         env,
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 30 |         execArgv: [
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 31 |             ...process.execArgv,
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 32 |             // If the worker is typescript, we need to register ts-node
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 33 |             ...(isWorkerTypescript ? ['-r', 'ts-node/register'] : []),
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 34 |         ],
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 35 |     });
     | ^^^^^^^
  36 |     worker.send({ type: 'load', payload: { plugin, root } });
  37 |     // logger.verbose(`[plugin-worker] started worker: ${worker.pid}`);
  38 |     const pendingPromises = new Map();



https://nextjs.org/docs/messages/module-not-found



./node_modules/nx/src/tasks-runner
Module not found: Can't resolve '/ROOT/node_modules/nx/src/tasks-runner/batch/run-batch.js'
server relative imports are not implemented yet. Please try an import relative to the file you are importing from.


https://nextjs.org/docs/messages/module-not-found



./node_modules/nx/src/tasks-runner
Module not found: Can't resolve '/ROOT/node_modules/nx/src/tasks-runner/remove-old-cache-records.js'
server relative imports are not implemented yet. Please try an import relative to the file you are importing from.


https://nextjs.org/docs/messages/module-not-found



./node_modules/@nx/devkit/src/utils/convert-nx-executor.js:51:12
Module not found: Can't resolve '@angular-devkit/architect'
  49 |         return toObservable(promise());
  50 |     };
> 51 |     return require('@angular-devkit/architect').createBuilder(builderFunction);
     |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  52 | }
  53 | exports.convertNxExecutor = convertNxExecutor;
  54 | function toObservable(promiseOrAsyncIterator) {



https://nextjs.org/docs/messages/module-not-found



./node_modules/nx/src/adapter/ngcli-adapter.js:109:27
Module not found: Can't resolve '@angular-devkit/architect'
  107 | exports.createBuilderContext = createBuilderContext;
  108 | async function scheduleTarget(root, opts, verbose) {
> 109 |     const { Architect } = require('@angular-devkit/architect');
      |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  110 |     const logger = (0, exports.getLogger)(verbose);
  111 |     const fsHost = new NxScopedHostForBuilders(root);
  112 |     const { workspace } = await core_1.workspaces.readWorkspace('angular.json', core_1.workspaces.createWorkspaceHost(fsHost));



https://nextjs.org/docs/messages/module-not-found



./node_modules/nx/src/adapter/ngcli-adapter.js:38:27
Module not found: Can't resolve '@angular-devkit/architect'
  36 |         return [];
  37 |     });
> 38 |     const { Architect } = require('@angular-devkit/architect');
     |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  39 |     const architect = new Architect(architectHost, registry);
  40 |     const { firstValueFrom } = require('rxjs');
  41 |     const toPromise = (obs) => firstValueFrom ? firstValueFrom(obs) : obs.toPromise();



https://nextjs.org/docs/messages/module-not-found



./node_modules/nx/src/adapter/ngcli-adapter.js:782:129
Module not found: Can't resolve '@angular-devkit/architect/node'
  780 | }
  781 | async function getWrappedWorkspaceNodeModulesArchitectHost(workspace, root, projects) {
> 782 |     const { WorkspaceNodeModulesArchitectHost: AngularWorkspaceNodeModulesArchitectHost, } = await Promise.resolve().then(() => require('@angular-devkit/architect/node'));
      |                                                                                                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  783 |     class WrappedWorkspaceNodeModulesArchitectHost extends AngularWorkspaceNodeModulesArchitectHost {
  784 |         constructor(workspace, root, projects) {
  785 |             super(workspace, root);



https://nextjs.org/docs/messages/module-not-found



./node_modules/nx/src/adapter/ngcli-adapter.js:4:16
Module not found: Can't resolve '@angular-devkit/core'
  2 | Object.defineProperty(exports, "__esModule", { value: true });
  3 | exports.getLogger = exports.wrapAngularDevkitSchematic = exports.mockSchematicsForTesting = exports.runMigration = exports.generate = exports.NxScopeHostUsedForWrappedSchematics = exports.arrayBufferToString = exports.NxScopedHostForBuilders = exports.NxScopedHost = exports.scheduleTarget = exports.createBuilderContext = void 0;
> 4 | const core_1 = require("@angular-devkit/core");
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  5 | const node_1 = require("@angular-devkit/core/node");
  6 | const chalk = require("chalk");
  7 | const path_1 = require("path");



https://nextjs.org/docs/messages/module-not-found



./node_modules/nx/src/adapter/ngcli-adapter.js:29:28
Module not found: Can't resolve '@angular-devkit/core'
  27 |     // the top level import is not patched because it is imported before the
  28 |     // patching happens so we require it here to use the patched version below
> 29 |     const { workspaces } = require('@angular-devkit/core');
     |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  30 |     const { workspace } = await workspaces.readWorkspace('angular.json', workspaces.createWorkspaceHost(fsHost));
  31 |     const architectHost = await getWrappedWorkspaceNodeModulesArchitectHost(workspace, context.root, context.projectsConfigurations.projects);
  32 |     const registry = new core_1.schema.CoreSchemaRegistry();



https://nextjs.org/docs/messages/module-not-found



./node_modules/nx/src/adapter/ngcli-adapter.js:5:16
Module not found: Can't resolve '@angular-devkit/core/node'
  3 | exports.getLogger = exports.wrapAngularDevkitSchematic = exports.mockSchematicsForTesting = exports.runMigration = exports.generate = exports.NxScopeHostUsedForWrappedSchematics = exports.arrayBufferToString = exports.NxScopedHostForBuilders = exports.NxScopedHost = exports.scheduleTarget = exports.createBuilderContext = void 0;
  4 | const core_1 = require("@angular-devkit/core");
> 5 | const node_1 = require("@angular-devkit/core/node");
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  6 | const chalk = require("chalk");
  7 | const path_1 = require("path");
  8 | const rxjs_1 = require("rxjs");



https://nextjs.org/docs/messages/module-not-found



./node_modules/nx/src/adapter/ngcli-adapter.js:171:56
Module not found: Can't resolve '@angular-devkit/schematics'
  169 |         packageManager: (0, package_manager_1.detectPackageManager)(),
  170 |         root: (0, core_1.normalize)(root),
> 171 |         registry: new core_1.schema.CoreSchemaRegistry(require('@angular-devkit/schematics').formats.standardFormats),
      |                                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  172 |         resolvePaths: [process.cwd(), root],
  173 |         engineHostCreator: (options) => createNodeModulesEngineHost(options.resolvePaths, projects),
  174 |     });



https://nextjs.org/docs/messages/module-not-found



./node_modules/nx/src/adapter/ngcli-adapter.js:176:50
Module not found: Can't resolve '@angular-devkit/schematics/tools'
  174 |     });
  175 |     workflow.registry.addPostTransform(core_1.schema.transforms.addUndefinedDefaults);
> 176 |     workflow.engineHost.registerOptionsTransform(require('@angular-devkit/schematics/tools').validateOptionsWithSchema(workflow.registry));
      |                                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  177 |     if (opts.interactive) {
  178 |         workflow.registry.usePromptProvider(createPromptProvider());
  179 |     }



https://nextjs.org/docs/messages/module-not-found



./node_modules/nx/src/adapter/ngcli-adapter.js:165:26
Module not found: Can't resolve '@angular-devkit/schematics/tools'
  163 | }
  164 | function createWorkflow(fsHost, root, opts, projects) {
> 165 |     const NodeWorkflow = require('@angular-devkit/schematics/tools').NodeWorkflow;
      |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  166 |     const workflow = new NodeWorkflow(fsHost, {
  167 |         force: false,
  168 |         dryRun: opts.dryRun,



https://nextjs.org/docs/messages/module-not-found



./node_modules/nx/src/adapter/ngcli-adapter.js:134:35
Module not found: Can't resolve '@angular-devkit/schematics/tools'
  132 | exports.scheduleTarget = scheduleTarget;
  133 | function createNodeModulesEngineHost(resolvePaths, projects) {
> 134 |     const NodeModulesEngineHost = require('@angular-devkit/schematics/tools')
      |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  135 |         .NodeModulesEngineHost;
  136 |     class NxNodeModulesEngineHost extends NodeModulesEngineHost {
  137 |         constructor() {



https://nextjs.org/docs/messages/module-not-found



./node_modules/@swc/core/index.js:68:28
Module not found: Can't resolve '@swc/wasm'
  66 |     catch (_) {
  67 |         // postinstall supposed to install `@swc/wasm` already
> 68 |         fallbackBindings = require("@swc/wasm");
     |                            ^^^^^^^^^^^^^^^^^^^^
  69 |     }
  70 |     finally {
  71 |         return binding;



https://nextjs.org/docs/messages/module-not-found



./node_modules/prettier/third-party.js:83:34
Module not found
  81 |       delete require.cache[filePath];
  82 |       const parent = require.cache[parentPath];
> 83 |       return parent === void 0 ? require(filePath) : parent.require(filePath);
     |                                  ^^^^^^^^^^^^^^^^^
  84 |     };
  85 |   }
  86 | });



https://nextjs.org/docs/messages/module-not-found



./node_modules/prettier/index.js:38143:10
Module not found
  38141 |       const externalPlugins = [...uniqByKey([...externalManualLoadPluginInfos, ...externalAutoLoadPluginInfos], "requirePath").map((externalPluginInfo) => Object.assign({
  38142 |         name: externalPluginInfo.name
> 38143 |       }, require(externalPluginInfo.requirePath))), ...externalPluginInstances];
        |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  38144 |       return [...internalPlugins, ...externalPlugins];
  38145 |     }
  38146 |     function findPluginsInNodeModules(nodeModulesDir) {



https://nextjs.org/docs/messages/module-not-found



./node_modules/nx/src/adapter/ngcli-adapter.js:588:26
Module not found
  586 |     }
  587 |     if ((0, path_1.extname)(name)) {
> 588 |         collectionPath = require.resolve(name);
      |                          ^^^^^^^^^^^^^^^^^^^^^
  589 |     }
  590 |     else {
  591 |         const { path: packageJsonPath, packageJson } = (0, package_json_1.readModulePackageJson)(name, (0, installation_directory_1.getNxRequirePaths)(process.cwd()));



https://nextjs.org/docs/messages/module-not-found



./node_modules/yargs-parser/build/index.cjs:1029:20
Module not found
  1027 |     require: (path) => {
  1028 |         if (typeof require !== 'undefined') {
> 1029 |             return require(path);
       |                    ^^^^^^^^^^^^^
  1030 |         }
  1031 |         else if (path.match(/\.json$/)) {
  1032 |             return JSON.parse(fs.readFileSync(path, 'utf8'));



https://nextjs.org/docs/messages/module-not-found



./node_modules/nx/src/adapter/ngcli-adapter.js:153:38
Module not found
  151 |                     throw new Error(`The "${name}" package does not support Nx generators or Angular Devkit schematics.`);
  152 |                 }
> 153 |                 collectionFilePath = require.resolve((0, path_1.join)((0, path_1.dirname)(packageJsonPath), schematics ?? generators));
      |                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  154 |             }
  155 |             return collectionFilePath;
  156 |         }



https://nextjs.org/docs/messages/module-not-found



./node_modules/nx/src/native/index.js:53:28
Module not found
  51 |     localNodeFiles.some((f) => modulePath.endsWith(f))
  52 |   ) {
> 53 |     const nativeLocation = require.resolve(modulePath);
     |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  54 |     const fileName = basename(nativeLocation);
  55 |
  56 |     // we copy the file to a workspace-scoped tmp directory and prefix with nxVersion to avoid stale files being loaded



https://nextjs.org/docs/messages/module-not-found



./node_modules/tsconfig-paths/lib/filesystem.js:31:12
Module not found
  29 |     }
  30 |     // eslint-disable-next-line @typescript-eslint/no-require-imports
> 31 |     return require(packageJsonPath);
     |            ^^^^^^^^^^^^^^^^^^^^^^^^
  32 | }
  33 | exports.readJsonFromDiskSync = readJsonFromDiskSync;
  34 | function readJsonFromDiskAsync(path,



https://nextjs.org/docs/messages/module-not-found



./node_modules/ts-node/dist-raw/node-internal-modules-cjs-helpers.js:67:21
Module not found
  65 |       get: () => {
  66 |         // Node 12 hack; remove when we drop node12 support
> 67 |         const lib = (dummyModule.require || require)(name);
     |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  68 |
  69 |         // Disable the current getter/setter and set up a new
  70 |         // non-enumerable property.



https://nextjs.org/docs/messages/module-not-found



./node_modules/nx/src/adapter/ngcli-adapter.js:808:39
Module not found
  806 |                 throw new Error(`The "${nodeModule}" package does not support Nx executors or Angular Devkit Builders.`);
  807 |             }
> 808 |             const executorsFilePath = require.resolve((0, path_1.join)((0, path_1.dirname)(packageJsonPath), executorsFile));
      |                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  809 |             const executorsJson = (0, fileutils_1.readJsonFile)(executorsFilePath);
  810 |             const executorConfig = executorsJson.builders?.[builder] ?? executorsJson.executors?.[builder];
  811 |             if (!executorConfig) {



https://nextjs.org/docs/messages/module-not-found



./node_modules/nx/src/command-line/run/executor-utils.js:65:31
Module not found
  63 |         throw new Error(`The "${nodeModule}" package does not support Nx executors.`);
  64 |     }
> 65 |     const executorsFilePath = require.resolve((0, path_1.join)((0, path_1.dirname)(packageJsonPath), executorsFile));
     |                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  66 |     const executorsJson = (0, fileutils_1.readJsonFile)(executorsFilePath);
  67 |     const executorConfig = executorsJson.executors?.[executor] || executorsJson.builders?.[executor];
  68 |     if (!executorConfig) {



https://nextjs.org/docs/messages/module-not-found



./node_modules/nx/src/project-graph/plugins/loader.js:212:64
Module not found
  210 | exports.loadNxPluginAsync = loadNxPluginAsync;
  211 | async function importPluginModule(pluginPath) {
> 212 |     const m = await Promise.resolve(`${pluginPath}`).then(s => require(s));
      |                                                                ^^^^^^^^^^
  213 |     if (m.default &&
  214 |         ('createNodes' in m.default || 'createDependencies' in m.default)) {
  215 |         return m.default;



https://nextjs.org/docs/messages/module-not-found



./node_modules/prettier/index.js:18428:31
Module not found
  18426 |                 paths: [dir]
  18427 |               });
> 18428 |               result.config = require(modulePath);
        |                               ^^^^^^^^^^^^^^^^^^^
  18429 |             }
  18430 |             if (typeof result.config !== "object") {
  18431 |               throw new TypeError(`Config is only allowed to be an object, but received ${typeof result.config} in "${result.filepath}"`);



https://nextjs.org/docs/messages/module-not-found



./node_modules/ts-node/dist/transpilers/swc.js:12:23
Module not found
  10 |     if (typeof swc === 'string') {
  11 |         swcDepName = swc;
> 12 |         swcInstance = require(transpilerConfigLocalResolveHelper(swc, true));
     |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  13 |     }
  14 |     else if (swc == null) {
  15 |         let swcResolved;



https://nextjs.org/docs/messages/module-not-found



./node_modules/ts-node/dist/index.js:247:39
Module not found
  245 |                 : projectLocalResolveHelper;
  246 |             const transpilerPath = transpilerConfigLocalResolveHelper(transpilerName, true);
> 247 |             const transpilerFactory = require(transpilerPath)
      |                                       ^^^^^^^^^^^^^^^^^^^^^^^
  248 |                 .create;
  249 |             return createTranspiler;
  250 |             function createTranspiler(compilerOptions, nodeModuleEmitKind) {



https://nextjs.org/docs/messages/module-not-found



./node_modules/ts-node/dist/transpilers/swc.js:29:23
Module not found
  27 |             }
  28 |         }
> 29 |         swcInstance = require(swcResolved);
     |                       ^^^^^^^^^^^^^^^^^^^^
  30 |     }
  31 |     else {
  32 |         swcInstance = swc;



https://nextjs.org/docs/messages/module-not-found



./node_modules/@nx/devkit/src/utils/package-json.js:370:12
Module not found
  368 | }
  369 | function getPackageVersion(pkg) {
> 370 |     return require((0, path_1.join)(pkg, 'package.json')).version;
      |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  371 | }
  372 | /**
  373 |  * @description The version of Nx used by the workspace. Returns null if no version is found.



https://nextjs.org/docs/messages/module-not-found



./node_modules/nx/src/config/schema-utils.js:20:24
Module not found
  18 |             (0, plugins_1.registerPluginTSTranspiler)();
  19 |         }
> 20 |         const module = require(modulePath);
     |                        ^^^^^^^^^^^^^^^^^^^
  21 |         return implementationExportName
  22 |             ? module[implementationExportName]
  23 |             : module.default ?? module;



https://nextjs.org/docs/messages/module-not-found


    at processIssues (/home/yuval/projects/nx-turbo/node_modules/next/dist/server/dev/turbopack-utils.js:229:15)
    at handleRouteType (/home/yuval/projects/nx-turbo/node_modules/next/dist/server/dev/turbopack-utils.js:386:17)
    at async Object.ensurePage (/home/yuval/projects/nx-turbo/node_modules/next/dist/server/dev/hot-reloader-turbopack.js:644:17)
    at async DevBundlerService.ensurePage (/home/yuval/projects/nx-turbo/node_modules/next/dist/server/lib/dev-bundler-service.js:18:20)
    at async DevServer.ensurePage (/home/yuval/projects/nx-turbo/node_modules/next/dist/server/dev/next-dev-server.js:551:9)
    at async Object.ensure (/home/yuval/projects/nx-turbo/node_modules/next/dist/server/dev/next-dev-server.js:169:17)
    at async DevRouteMatcherManager.matchAll (/home/yuval/projects/nx-turbo/node_modules/next/dist/server/future/route-matcher-managers/dev-route-matcher-manager.js:96:13)
    at async DevRouteMatcherManager.match (/home/yuval/projects/nx-turbo/node_modules/next/dist/server/future/route-matcher-managers/default-route-matcher-manager.js:155:26)
    at async NextNodeServer.handleCatchallRenderRequest (/home/yuval/projects/nx-turbo/node_modules/next/dist/server/next-server.js:228:31)
    at async DevServer.handleRequestImpl (/home/yuval/projects/nx-turbo/node_modules/next/dist/server/base-server.js:816:17)
    at async /home/yuval/projects/nx-turbo/node_modules/next/dist/server/dev/next-dev-server.js:339:20
    at async Span.traceAsyncFn (/home/yuval/projects/nx-turbo/node_modules/next/dist/trace/trace.js:154:20)
    at async DevServer.handleRequest (/home/yuval/projects/nx-turbo/node_modules/next/dist/server/dev/next-dev-server.js:336:24)
    at async invokeRender (/home/yuval/projects/nx-turbo/node_modules/next/dist/server/lib/router-server.js:174:21)
    at async handleRequest (/home/yuval/projects/nx-turbo/node_modules/next/dist/server/lib/router-server.js:353:24)
    at async requestHandlerImpl (/home/yuval/projects/nx-turbo/node_modules/next/dist/server/lib/router-server.js:377:13)
    at async Server.requestListener (/home/yuval/projects/nx-turbo/node_modules/next/dist/server/lib/start-server.js:141:13)
 ○ Compiling /_error ...
 ✓ Compiled /_error in 1979ms
 GET / 500 in 32076ms
```

### Operating System

- [ ] macOS
- [X] Linux
- [X] Windows
- [ ] Other (Please specify)

### Additional Information

Same result using both WSL Fedora Remix 39.01 and Windows

**Comment by fernandoMendozaMM** [comment]
I'm also facing the same issue, not even sure what caused it because I think it was working fine before.

**Comment by cyrus-za** [comment]
Also facing this. Not sure why we even got angular devkit in the mix as there's no angular code in the repo. I guess nx is using some code from angular to handle its monorepo orchestration

**Comment by ndcunningham** [comment]
Hi, I think this is an issue with Next.js / turbopack and not necessarily Nx. 
I opened an issue on Next.js GH about it: https://github.com/vercel/next.js/issues/66602

So we have to wait for the Vercel team to address it.

**Comment by mxa0079** [comment]
Thanks @ndcunningham - however the issue you linked seems to be purely related to Tailwind.

In this case the logs are showing errors importing Angular modules within a nx adaptar, for a repo that does not have any angular projects in it. 

So is this a different issue?

Is turbopack not working for NX at the moment? Do we know which one was the last version on which it was supported? Is there a known workaround?

**Comment by jgalbsss** [comment]
bumping this, same issue

**Comment by duarte2377** [comment]
same issue with next-i18next package

**Comment by DrewLandgraveCbsi** [comment]
Just ran into this as well after upgrading to NextJs 15 where Turbo is supposed to be stable for dev

**Comment by beany-vu** [comment]
can I ask if the issue is solved for now?

**Comment by nurulhudaapon** [comment]
> can I ask if the issue is solved for now?

Not working for us yet.

**Comment by kustomq** [comment]
Same issue here, next v15 & nx v20

```
hook.js:608 ./node_modules/.pnpm/nx@20.0.3_@swc-node+register@1.9.2_@swc+core@1.5.7_@swc+helpers@0.5.11__@swc+types@0.1.7_type_cxohwuiq2lf5alcyvrfldxcl4e/node_modules/nx/src/tasks-runner
Module not found: Can't resolve '/ROOT/node_modules/.pnpm/nx@20.0.3_@swc-node+register@1.9.2_@swc+core@1.5.7_@swc+helpers@0.5.11__@swc+types@0.1.7_type_cxohwuiq2lf5alcyvrfldxcl4e/node_modules/nx/src/tasks-runner/batch/run-batch.js'
server relative imports are not implemented yet. Please try an import relative to the file you are importing from.
```

![Image](https://github.com/user-attachments/assets/b94f128d-ebf5-42d8-9214-84046a2262af)

**Comment by arvilmena** [comment]
So the summary is that: https://github.com/vercel/next.js/issues/60363#issuecomment-1890480269
which basically means NX doesn't play nicely with CJS at the moment

Is there a plan to resolve this? If so, I propose removing the `blocked:third-party` label since we cannot blame the NextJS nor tailwind on this anymore.
Not resolving this means future NextJS users cannot consider NX as viable option since the `--turbo` that speeds up development is broken in `NX`

@FrozenPandaz  @jaysoo @ndcunningham

**Comment by KostiantynPopovych** [comment]
+ Highly waiting for any final resolution

**Comment by jgalbsss** [comment]
Any updates, seems to be a common issue

**Comment by usman-nasir-kwanso** [comment]
Getting Same issue with nx,
Any updates, seems to be a common issue
![Image](https://github.com/user-attachments/assets/51e178d1-4d8e-4c06-b838-45a97b3a736b)

**Comment by nurulhudaapon** [comment]
Just want to share that we were able to resolve that on a relatively large project, it was not related NX even though the error message suggests that. For us it was mostly about some commonjs module, and we were able to resolve those after replacing. The way we debug was removing almost all dependancies from our project and then adding page by page to see which is bringing up the error.

**Comment by jamiter** [comment]
@nurulhudaapon, very interesting! Would you mind sharing which packaged caused the issues for you?

**Comment by DreamwareDevelopment** [comment]
Also experiencing this issue. Next's dev webpack is so slow, please help

**Comment by tonipepperoni** [comment]
issue related with createGlobPatternforDependencies() - it appears turbopack doesn't like the cjs running on that function for tailwind config

temp fix: console log output of createGlobPatternforDependencies, then paste the dependencies manually.

**Comment by tonipepperoni** [comment]
alternatively run tw compiler: 
https://tailwindcss.com/docs/installation

**Comment by gentle-noah** [comment]
Bumping this as the inability to use turbopack is now dramatically hindering my team's local DX. We're started entertaining migrating our repo over to truborepo so we can experience all of the modern benefits of NextJS.

**Comment by marko-hologram** [comment]
> issue related with createGlobPatternforDependencies() - it appears turbopack doesn't like the cjs running on that function for tailwind config
> 
> temp fix: console log output of createGlobPatternforDependencies, then paste the dependencies manually.

Thank you for this. This let me narrow down the issue, but I ran into another issue then with these globs not working for edits to styles. New classes weren't being applied unless I stopped the app and started it again. No matter if I did updates to files inside the app directly or packages that were outside the app.

I opened an issue about it here https://github.com/nrwl/nx/issues/29946

**Comment by ndcunningham** [comment]
Hey sorry for the lack of updates.

From the OP the issue here is with Next.js `--turbo` and how it handles CJS imports.
If `--turbo` is not used meaning bundling with Webpack the default configuration generated by Nx works. There should be feature parity between Next.js and Webpack but this is not the case at the moment.

Our proposed way forward would be by default to not include `createGlobPatternsForDependencies`.
This will be generated but commented out such that if you are not using `--turbo` you can include them by uncommenting the statements and it works out of the box.

Additionally, @marko-hologram did a great write-up of an additional issue related to pathing. All these changes will be included in https://github.com/nrwl/nx/pull/30192

**Comment by pciruzzi** [comment]
@ndcunningham I've updated my app to use NX 20.6.0, which should include the fix, but I'm still facing the issue mentioned here. I'm using Next 15.2.2 and Node 20. What could be wrong? Thanks!

**Comment by marko-hologram** [comment]
> [@ndcunningham](https://github.com/ndcunningham) I've updated my app to use NX 20.6.0, which should include the fix, but I'm still facing the issue mentioned here. I'm using Next 15.2.2 and Node 20. What could be wrong? Thanks!

Fix is to actually not use `createGlobPatternsForDependencies` directly when using Turbopack. If that function is being used in your tailwind.config.js file, that's causing issues.

I'm guessing updating NX doesn't update existing Tailwind config files and the fix is only being added to newly generated files. You can re-read this whole discussion for potential ways to fix this manually.

**Comment by pciruzzi** [comment]
> > [@ndcunningham](https://github.com/ndcunningham) I've updated my app to use NX 20.6.0, which should include the fix, but I'm still facing the issue mentioned here. I'm using Next 15.2.2 and Node 20. What could be wrong? Thanks!
> 
> Fix is to actually not use `createGlobPatternsForDependencies` directly when using Turbopack. If that function is being used in your tailwind.config.js file, that's causing issues.
> 
> I'm guessing updating NX doesn't update existing Tailwind config files and the fix is only being added to newly generated files. You can re-read this whole discussion for potential ways to fix this manually.

Hi @marko-hologram ! First of all, thanks for the reply. That being said, I thought that #30192 fixed that in particular and I didn't deep dive on the changes, but now I understand better. I will try to make the corresponding changes and see if it works with those. Thanks 😄

**Comment by Playjasb2** [comment]
Any update on this issue? I have one project on Nx v20.5.0 and were blocked on using Turbopack. I have an older project on v19.6.1 that also has this problem as well, which proves that this is a very long standing issue.

We are literally blocked from having faster or better tooling for development for at least 8 months or so.

When can we get a fix for this?

**Comment by doctorloaf** [comment]
> Any update on this issue? I have one project on Nx v20.5.0 and were blocked on using Turbopack. I have an older project on v19.6.1 that also has this problem as well, which proves that this is a very long standing issue.
> 
> We are literally blocked from having faster or better tooling for development for at least 8 months or so.
> 
> When can we get a fix for this?

not trying to pull people away from NX, but I will say our migration to turborepo took maybe 1.5 days of annoying work and we were deployed on vercel at the end of day 2. happy to chat about it if you decide to take the leap. Our initial compile times went from ~5minutes to <60s and our page refreshes went from ~1-3m to <15s.

**Comment by marko-hologram** [comment]
> Any update on this issue? I have one project on Nx v20.5.0 and were blocked on using Turbopack. I have an older project on v19.6.1 that also has this problem as well, which proves that this is a very long standing issue.
> 
> We are literally blocked from having faster or better tooling for development for at least 8 months or so.
> 
> When can we get a fix for this?

If this is the only thing blocking you from adopting Turbopack with NX, just follow instructions from all the related comments and issues. Basically, remove `createGlobPatternsForDependencies` function from Tailwind config(s) and just write globs manually. You can most likely even target everything in "packages" folder since Tailwind content scanning is pretty fast as far as I know.

For me it was easy to do because each app in my repo has a dedicated folder inside "packages" so I could target just them. For example, if my app is called "basic", I'll have "packages/basic/authentication" as a package and I was able to just target "packages/basic/**" for files to scan with Tailwind. I have some shared content as well, so my Tailwind content config looks something like this:

```js
  content: [
    // Target local files inside the app
    "./src/**/*.{js,ts,jsx,tsx,mdx}",
    // Target shared component library
    "./../../packages/components/**/*.{ts,tsx,js,jsx,html}",
    // Target packages belonging to the app
    "./../../packages/basic/**/*.{ts,tsx,js,jsx,html}",
  ],
```

Running the app with Turbopack works, hot reload works etc.

**Comment by dlhck** [comment]
@Playjasb2 these two steps are necessary to get it to work: 

1. do what @marko-hologram has outline above
2. set the `outputFileTracingRoot` in your `next.config.js` (optional, try if it works without first)

```
if (process.env.TURBOPACK === 'true') {
  nextConfig.outputFileTracingRoot = path.join(__dirname, '../../')
}
```

**Comment by Playjasb2** [comment]
@doctorloaf I did look into this. It's just that it will take time and effort to switch over. Plus we were used to certain workflows.

@marko-hologram, @dlhck Ah thank you! I guess I misread or didn't find the solution obvious at the first when I was going through it the other day. I managed to get this to work. I just think like out-of-the-box, the setup shouldn't break when we bring in Tailwind while Turbopack is enabled. It took us having to go to this particular issue to get a solution.

**Comment by bbyiringiro** [comment]
The main issue when using Next.js with Turbopack in an Nx monorepo involves createGlobPatternsForDependencies from @nx/react/tailwind. Turbopack doesn't handle dynamic imports and CommonJS resolution patterns used by this function, causing module resolution errors. To fix this:

`// Before (doesn't work with Turbopack)
content: [
  join(__dirname, '{src,pages,app}/**/*.{ts,tsx,html}'),
  ...createGlobPatternsForDependencies(__dirname)
],

// After (Turbopack compatible)
content: [
  join(__dirname, '{src,pages,app}/**/*.{ts,tsx,html}'),
  join(__dirname, '../../libs/ui-components/src/**/*.{ts,tsx}'),
  join(__dirname, '../../libs/shared/src/components/**/*.{ts,tsx}')
],`


A related issue is that Turbopack doesn't support CommonJS libraries well. You may need to convert libraries with type "commonjs" to "module" in their package.json files (particularly in shared libraries).

**Comment by azharuddin-mulla-amla** [comment]
when using `--turbo`, In nextjs node runtime not supported, It edge runtime support so we can't use commonJs libraries like `require, fs, path module`.

**Comment by mxa0079** [comment]
Indeed, the main problem is `createGlobPatternsForDependencies`, but in my case, I first had to migrate to .mjs config files.

Here’s how I was able to make it work:

⸻

### First Step: Migrate to .mjs Files

First, I ensured all my configuration files (Tailwind and PostCSS) were defined as .mjs instead of .cjs.

This required me to:

1. Change the file extensions from .cjs to .mjs
2. Add a helper function to define  __dirname (which is not available in .mjs files): `const __dirname = dirname(fileURLToPath(import.meta.url));`
3. Change `module.exports =` to `export default`
4. Fix tailwind.config.mjs to ensure the plugins property is an array (i.e., change `plugins: tailwindcssAnimate`, to `plugins: [tailwindcssAnimate]` -- notice the use of an array)

Once that was working (without enabling Turbopack yet), it was time to remove the usage of `createGlobPatternsForDependencies`.

⸻

### Second Step: Remove `createGlobPatternsForDependencies`

Getting rid of `createGlobPatternsForDependencies` was a bit more involved than I initially thought, mainly because the project has dependencies on other nested projects within the monorepo. Here are the steps I followed:

1. Print out the project dependencies detected by createGlobPatternsForDependencies:

	

``` ts
console.log(
  `GLOBPATTERNS output: ${JSON.stringify(
    createGlobPatternsForDependencies(__dirname)
  )}`
);
```


2. Replace the problematic call with hardcoded dependency paths (i.e., the output from the previous step) and ensure everything works. At this point, you can enable Turbopack and it should work. However, this introduces a strong dependency on your local filesystem structure, which leads to the next step.
3. Create the following helper function to correctly resolve the absolute paths of your files:


```ts
import { dirname, join, resolve } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));

function getLibGlobs() {
  const root = resolve(__dirname, '../..'); // adjust to root of monorepo
  const libDirs = [
    'libs/shadcn-ui/src',
    'libs/dora-sdk/src',
    'libs/portal/user-mgmt/src',
    'libs/wl-prisma-schema/src',
    'libs/catalog-prisma-schema/src',
  ];

  return libDirs.map((lib) =>
    join(root, lib, '**/*!(*.stories|*.spec).{ts,tsx,html}') // NOTE: If you are using js or jsx, add those extensions.
  );
}

```

4. Replace the previously hardcoded paths (or the call to createGlobPatternsForDependencies) with a call to the getLibGlobs function:


```ts
content: [
    join(
      __dirname,
      '{src,pages,components,app}/**/*!(*.stories|*.spec).{ts,tsx,html}'
    ),
    ...getLibGlobs(),
  ],
```




⸻

And that’s it!

Limitation: Any new library dependencies in your Next.js project will need to be manually added to the libDirs array. I suppose you could write a function to detect these dependencies dynamically, but this solution is good enough for me.

Hope this helps!

**Comment by Coly010** [comment]
This should be solved in the latest version of Nx where we added support for Next 16. We identified an issue with how turbopack handles the next.config that caused issues with the Nx Daemon
